pkg/webhook/cvc.go

```diff
@@ -0,0 +1,199 @@
+	cvc "github.com/openebs/maya/pkg/cstorvolumeclaim/v1alpha1"
+	util "github.com/openebs/maya/pkg/util"
+	"github.com/pkg/errors"
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+)
+
+type validateFunc func(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error
+
+type getCVC func(name, namespace string, clientset clientset.Interface) (*apis.CStorVolumeClaim, error)
+
+func (wh *webhook) validateCVCUpdateRequest(req *v1beta1.AdmissionRequest, getCVC getCVC) *v1beta1.AdmissionResponse {
+	response := NewAdmissionResponse().
+		SetAllowed().
+		WithResultAsSuccess(http.StatusAccepted).AR
+	var cvcNewObj apis.CStorVolumeClaim
+	err := json.Unmarshal(req.Object.Raw, &cvcNewObj)
+	if err != nil {
+		klog.Errorf("Couldn't unmarshal raw object: %v to cvc error: %v", req.Object.Raw, err)
+		response = BuildForAPIObject(response).UnSetAllowed().WithResultAsFailure(err, http.StatusBadRequest).AR
+		return response
+	}
+
+	// Get old CVC object by making call to etcd
+	cvcOldObj, err := getCVC(cvcNewObj.Name, cvcNewObj.Namespace, wh.clientset)
+	if err != nil {
+		klog.Errorf("Failed to get CVC %s in namespace %s from etcd error: %v", cvcNewObj.Name, cvcNewObj.Namespace, err)
+		response = BuildForAPIObject(response).UnSetAllowed().WithResultAsFailure(err, http.StatusBadRequest).AR
+		return response
+	}
+	err = validateCVCSpecChanges(cvcOldObj, &cvcNewObj)
+	if err != nil {
+		klog.Errorf("invalid cvc changes: %s error: %s", cvcOldObj.Name, err.Error())
+		response = BuildForAPIObject(response).UnSetAllowed().WithResultAsFailure(err, http.StatusBadRequest).AR
+		return response
+	}
+	return response
+}
+
+func validateCVCSpecChanges(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	validateFuncList := []validateFunc{validateReplicaCount,
+		validatePoolListChanges,
+		validateReplicaScaling,
+		validateStatusPoolList}
+	for _, f := range validateFuncList {
+		err := f(cvcOldObj, cvcNewObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Below validations should be done only with new CVC object
+	err := validatePoolNames(cvcNewObj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// TODO: isScalingInProgress(cvcObj *apis.CStorVolumeClaim) signature need to be
+// updated to cvcObj.IsScaleingInProgress()
+func isScalingInProgress(cvcObj *apis.CStorVolumeClaim) bool {
+	return len(cvcObj.Spec.Policy.ReplicaPoolInfo) != len(cvcObj.Status.PoolInfo)
+}
+
+// validateReplicaCount returns error if user modified the replica count after
+// provisioning the volume else return nil
+func validateReplicaCount(cvcOldObj, cvcNewObj *apis.CStorVolumeClaim) error {
+	if cvcOldObj.Spec.ReplicaCount != cvcNewObj.Spec.ReplicaCount {
+		return errors.Errorf(
+			"cvc %s replicaCount got modified from %d to %d",
+			cvcOldObj.Name,
+			cvcOldObj.Spec.ReplicaCount,
+			cvcNewObj.Spec.ReplicaCount,
+		)
+	}
+	return nil
+}
+
+// validatePoolListChanges returns error if user modified only the pool names
```

Comment:
validatePoolListChanges returns error if user modified existing pool name(s) or changed the poolnames count more then 1.  At a time only 1 scale down operation has been supported, but replica scale up can be more then 1

Reply:
Updated